### PR TITLE
[plug-in] Implement 'plugins' API

### PR DESCRIPTION
**Pull request by evidolob**
Add namespace for dealing with installed plug-ins. Plug-ins are represented
by an Plugin-interface which enables reflection on them.
Plug-in writers can provide APIs to other plug-ins by returning their API public
API from the `start`-call.

Issue: #2561

Signed-off-by: Yevhen Vydolob <yvydolob@redhat.com>

### CODE REVIEW

File: packages/plugin-ext/src/hosted/browser/worker/worker-main.ts [hunk 1]
```diff
@@ -55,32 +53,49 @@ rpc.set(MAIN_RPC_CONTEXT.HOSTED_PLUGIN_MANAGER_EXT, new HostedPluginManagerExtIm
...
         }
     },
-    stopPlugins(contextPath: string, pluginIds: string[]): void {
-        pluginIds.forEach(pluginId => {
-            const pluginData = plugins.get(pluginId);
-            if (pluginData) {
-                // call stop method
-                if (pluginData.stopPluginMethod) {
-                    pluginData.stopPluginMethod();
+    init(raw: PluginMetadata[]): [Plugin[], Plugin[]] {
```
Comment (benoitf) [inline packages/plugin-ext/src/hosted/browser/worker/worker-main.ts]: should we use something more meaningful than raw if we know it's of type pluginMetadata ?

File: packages/plugin-ext/src/plugin/plugin-context.ts [hunk 2]
```diff
@@ -18,9 +18,8 @@ import { CommandRegistryImpl } from './command-registry';
 import { Emitter } from '@theia/core/lib/common/event';
 import { CancellationTokenSource } from '@theia/core/lib/common/cancellation';
 import { QuickOpenExtImpl } from './quick-open';
-import { MAIN_RPC_CONTEXT, Plugin } from '../api/plugin-api';
+import { MAIN_RPC_CONTEXT, Plugin as PPlugin, PluginManager } from '../api/plugin-api';
```
Comment (benoitf) [inline packages/plugin-ext/src/plugin/plugin-context.ts]: could we use another name than PPlugin which might not be clear in the rest of the code

File: packages/plugin-ext/src/api/plugin-api.ts [hunk 3]
```diff
@@ -32,18 +32,35 @@ import {
...
-    $stopPlugin(contextPath: string): PromiseLike<void>;
+export interface PluginInitData {
+    plugins: PluginMetadata[];
 }
 
 export interface Plugin {
     pluginPath: string;
+    initPath: string;
     model: PluginModel;
+    rowModel: PluginPackage;
```
Comment (akurinnoy) [inline packages/plugin-ext/src/api/plugin-api.ts]: typo? maybe `rawModel`?

File: packages/plugin-ext/src/common/plugin-protocol.ts [hunk 4]
```diff
@@ -360,7 +361,7 @@ export interface PluginLifecycle {
  * The export function of initialization module of backend plugin.
  */
 export interface BackendInitializationFn {
-    (rpc: RPCProtocol, pluginMetadata: PluginMetadata): void;
+    (rpc: RPCProtocol, manger: PluginManager, pluginMetadata: PluginMetadata): void;
```
Comment (akurinnoy) [inline packages/plugin-ext/src/common/plugin-protocol.ts]: typo `manger` -> `manager`

File: packages/plugin-ext/src/plugin/plugin-manager.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+export interface PluginHost {
+
+    // tslint:disable-next-line:no-any
+    loadPlugin(contextPath: string, plugin: Plugin): any;
+
+    init(data: PluginMetadata[]): [Plugin[], Plugin[]];
+}
+
+interface StopFun {
```
Comment (akurinnoy) [inline packages/plugin-ext/src/plugin/plugin-manager.ts]: If you don't mind let it be `StopFn` or `StopFunc`.

Comment (benoitf) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: I was thinking the same :-) let's still have some FUN :-)

**Comment by benoitf** [comment]
are you gonna remove the merge commit (use rebase instead ? )

**Comment by evidolob** [comment]
Yes, I wait approve before squashing commits

**Pull request merged**